src/Features/LanguageServer/Protocol/ExternalAccess/Razor/SemanticTokensRangesHandler.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[Method(SemanticRangesMethodName)]
+internal class SemanticTokensRangesHandler : ILspServiceRequestHandler<SemanticTokensRangesParams, SemanticTokens>
+{
+    private const int TokenSize = 5;
+    public const string SemanticRangesMethodName = "textDocument/semanticTokens/ranges";
+    private readonly IGlobalOptionService _globalOptions;
+    private readonly SemanticTokensRefreshQueue _semanticTokenRefreshQueue;
+
+    public SemanticTokensRangesHandler(IGlobalOptionService globalOptions, SemanticTokensRefreshQueue semanticTokensRefreshQueue)
+    {
+        _globalOptions = globalOptions;
+        _semanticTokenRefreshQueue = semanticTokensRefreshQueue;
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<SemanticTokens> HandleRequestAsync(
+            SemanticTokensRangesParams request,
+            RequestContext context,
+            CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(request.TextDocument, "TextDocument is null.");
+        if (request.Ranges.Length == 0)
+        {
+            return new SemanticTokens() { Data = Array.Empty<int>() };
+        }
+
+        var responseData = new List<int[]>();
+        foreach (var range in request.Ranges)
+        {
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = request.TextDocument,
+                Range = range,
+            };
+
+            var semanticTokens = await HandleRequestAsync(newParams, context, cancellationToken).ConfigureAwait(false);
+            if (semanticTokens is not null)
+            {
+                responseData.Add(semanticTokens.Data);
+            }
+        }
+
+        return new SemanticTokens() { Data = StitchSemanticTokenResponsesTogether(responseData) };
+    }
+
+    internal async Task<LSP.SemanticTokens> HandleRequestAsync(
```

Comment:
this method is commonly used between SemanticTokensRangesHandler.HandleRequestAsync` and `SemanticTokensRangeHandler.HandleRequestAsync`. Do we want to keep it back in `SemanticTokensRangeHandler` and just call it here? or is it worth moving here and cleaning up this file a bit further?